### PR TITLE
14 generate types from a graphql schema

### LINKED ISSUE #14: Generate types from a GraphQL schema
Get GraphQL schema types in React tu surpress query errors in IDE. It would be best if possible to generate from remote server using one command run on Next startup, maybe something like [this](https://stackoverflow.com/a/72694402)

### PR DESCRIPTION
**Pull request by rubikon02**
# Features

Type safety when using GraphQL

# Usage

1. Create `.graphql` file in `@graphql` directory
2. Run `npm run gql:generate` to generate once or `npm run gql:watch` to generate when changes are made
3. Use types from `@graphql/types`

## Server side example

```tsx
import {ProjectDocument} from "@graphql/types";

const { data } = await getClient().query({
    query: ProjectDocument,
    variables: {
        projectUuid: "abc..."
    },
});
```

### With props

```tsx
// ...
return <Project {...data.project!} />
```

```tsx
import {Project} from "@graphql/types";

function Project(props: Project) {
    // ...
}
```

## Client side example

```tsx
import {useQuery} from "@apollo/client";
import {ProjectDocument} from "@graphql/types";
import {Alert, LinearProgress} from "@mui/material";

const { loading, data, error } = useQuery(ProjectDocument, {
  variables: {
      projectUuid: props.uuid
  },
});
if (loading) return <LinearProgress />
if (error) return <Alert severity="error">{`${error}`}</Alert>
```

# Issues

Env variable `NEXT_PUBLIC_GQL_SERVER` cannot be used in `@graphql/codegen.ts` config:

```
× Invalid Codegen Configuration!                                                         
                                                                                         
        Please make sure that your codegen config file contains either the "schema" field
        or every generated file has its own "schema" field. 
```

### CODE REVIEW

File: app/project/[uuid]/page.tsx [hunk 1]
```diff
@@ -1,20 +1,20 @@
...
+        query: ProjectDocument,
         variables: {
             projectUuid: params.uuid
         },
     });
 
     return (
         <>
-            <Project {...data.project} />
+            <Project {...data.project!} />
```
Comment (Spookyless) [inline app/project/[uuid]/page.tsx]: No Non-Null Assertion operators allowed - check if object is truthy instead

**Review by Spookyless — CHANGES_REQUESTED** [review]
Add custom scalars to config:
https://stackoverflow.com/questions/74890201/how-to-add-custom-scalars-to-graphql-codegen

File: graphql/codegen.ts [hunk 2]
```diff
@@ -0,0 +1,18 @@
+import { CodegenConfig } from '@graphql-codegen/cli'
+
+const config: CodegenConfig = {
+    schema: process.env.NEXT_PUBLIC_GQL_SERVER,
```
Comment (Spookyless) [inline graphql/codegen.ts:7]: Add as dev dependency to load env vars during codegen:
https://www.npmjs.com/package/dotenv-cli

**Review by Spookyless — CHANGES_REQUESTED** [review]

File: graphql/scalars.graphql
Comment (Spookyless) [inline graphql/scalars.graphql:1]: Why delele this file? Now other graphql files will report that `UUID` is missing

**Review by Spookyless — CHANGES_REQUESTED** [review]

File: graphql/scalars.graphql
Comment (rubikon02) [reply graphql/scalars.graphql:1]: Graphql files don't report anything and types are generated

image

Comment (rubikon02) [reply graphql/scalars.graphql:1]: Timestamp was already used in project without definition in schema

Comment (Spookyless) [reply graphql/scalars.graphql:1]: Right, because schema is taken from api now

**Review by Spookyless — APPROVED** [review]

**Pull request merged**